src/nupic/research/frameworks/htm/temporal_memory/sequence_memory_apical_tiebreak.py

```diff
@@ -0,0 +1,196 @@
+# Copyright (C) 2022, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from nupic.research.frameworks.htm.temporal_memory import TemporalMemoryApicalTiebreak
+
+import torch
+
+real_type = torch.float32
+int_type = torch.int64
+
+device = "cuda" if torch.cuda.is_available() else "cpu"
+
+class SequenceMemoryApicalTiebreak(TemporalMemoryApicalTiebreak):
+    """
+    sequence memory with apical tiebreak, built on temporal memory with apical tiebreak.
+    """
+
+    def __init__(
+        self,
+        num_minicolumns=2048,
+        apical_input_size=0,
+        num_cells_per_minicolumn=32,
+        activation_threshold=13,
+        reduced_basal_threshold=13,
+        initial_permanence=0.21,
+        connected_permanence=0.50,
+        matching_threshold=10,
+        sample_size=20,
+        permanence_increment=0.1,
+        permanence_decrement=0.1,
+        basal_segment_incorrect_decrement=0.0,
+        apical_segment_incorrect_decrement=0.0,
+        max_synapses_per_segment=-1,
+        seed=42
+    ):
+
+        params = {
+            "num_minicolumns" : num_minicolumns,
+            "basal_input_size" : num_minicolumns * num_cells_per_minicolumn,
+            "apical_input_size" : apical_input_size,
+            "num_cells_per_minicolumn" : num_cells_per_minicolumn,
+            "activation_threshold" : activation_threshold,
+            "reduced_basal_threshold" : reduced_basal_threshold,
+            "initial_permanence" : initial_permanence,
+            "connected_permanence" : connected_permanence,
+            "matching_threshold" : matching_threshold,
+            "sample_size" : sample_size,
+            "permanence_increment" : permanence_increment,
+            "permanence_decrement" : permanence_decrement,
+            "basal_segment_incorrect_decrement" : basal_segment_incorrect_decrement,
+            "apical_segment_incorrect_decrement" : apical_segment_incorrect_decrement,
+            "max_synapses_per_segment" : max_synapses_per_segment,
+            "seed" : seed
+        }
+
+        super().__init__(**params)
+
+        self.previous_apical_input = torch.empty(0, dtype=int_type).to(device)
```

Comment:
If you are initializing a tensor, you can instead pass a device argument to the `empty` function. Any pytorch function that initializes a tensor will take both a dtype and a device argument.

Reply:
Will incorporate